# Разогрев

## Задание 1

Иногда, когда нужно создавать много однотипных лямбда функций, еще удобнее будет создать функцию, которая будет их генерировать.

Реализуйте функцию ``mod_checker(x, mod=0)``, которая будет **возвращать лямбда функцию** от одного аргумента ``y``, которая будет возвращать ``True``, если остаток от деления ``y`` на ``x`` равен ``mod``, и ``False`` иначе.

Если в ``return`` вашей функции написано выражение, отличное от лямбда функции - задание считается проваленным

**Sample:**

```python
mod_3 = mod_checker(3)

print(mod_3(3)) # True
print(mod_3(4)) # False

mod_3_1 = mod_checker(3, 1)
print(mod_3_1(4)) # True
```

In [17]:
def mod_checker(x: int, mod=0):
    """функция возвращает лямбда функцию от аргумента y,
    которая возвращает True, если остаток деления y на x равен mod"""
    return lambda y: y%x == mod

mod_3 = mod_checker(3)

print(mod_3(3)) # True
print(mod_3(4)) # False

mod_3_1 = mod_checker(3, 1)
print(mod_3_1(4)) # True



True
False
True


# Numpy

In [18]:
import numpy as np

## Задание 2
Напишите функцию, которая принимает один аргумент - ``n``. Функция должна создать нампаевскую матрицу размером ``(n, n)`` и залить ее шахматным узором. Порядок узора не имеет значения, важен общий паттерн.

**Sample:**
```python
chess_maze(8)

[[0 1 0 1 0 1 0 1]
 [1 0 1 0 1 0 1 0]
 [0 1 0 1 0 1 0 1]
 [1 0 1 0 1 0 1 0]
 [0 1 0 1 0 1 0 1]
 [1 0 1 0 1 0 1 0]
 [0 1 0 1 0 1 0 1]
 [1 0 1 0 1 0 1 0]]
```



In [19]:
def chess_maze(n: int) -> np.array:
    """функция принимает на вход целое число и возвращает
    матрицу n на n, заполненную в шахматном порядке"""
    matrix_ = np.zeros((n,n), dtype=int) # создание матрицы размером (n,n), заполнееной нулями

    for i in range(n):
        for j in range(n):
            if j%2 == 0 and i%2 == 0:
                matrix_[i,j] =1 # присваиваем четным элементам четных строк значение 1
            if j%2 != 0 and i%2 != 0:
                matrix_[i,j] =1 # присваиваем нечетным элементам нечетных строк значение 1

    return matrix_

chess_maze(7)


array([[1, 0, 1, 0, 1, 0, 1],
       [0, 1, 0, 1, 0, 1, 0],
       [1, 0, 1, 0, 1, 0, 1],
       [0, 1, 0, 1, 0, 1, 0],
       [1, 0, 1, 0, 1, 0, 1],
       [0, 1, 0, 1, 0, 1, 0],
       [1, 0, 1, 0, 1, 0, 1]])

## Задание 3
Напишите функцию, которая принимает 2 аргумента ``arr`` и ``exclude_values``. ``arr`` - это одномерный нампаевский массив, в котором необходимо поменять знак всем числам (инвертировать), которые есть в массиве ``exclude_values``. Действия необходимо делать прямо в исходном массиве, не создавая копий.

**Sample:**
```python
#Исходный вектор 
arr = np.arange(10+1)
exclude_values = np.arange(4, 8)
 
my_func(arr, exclude_values)
array([ 0  1  2  3 -4 -5 -6 -7  8  9 10])
```

In [20]:
from http.client import NETWORK_AUTHENTICATION_REQUIRED


def my_func(arr: np.array, ex_values: np.array) -> np.array:
    """функция принимает на вход два массива и возвращает первый массив,
    в котором числа, которые есть во втором массиве инвертированы"""
    for i in range(len(arr)):
        if (arr[i] == ex_values).sum() != 0: # проверка наличия i-го элемента в массиве ex_values
            arr[i] = - arr[i]

    return arr

arr = np.arange(10+1)
exclude_values = np.arange(4, 8)
 
my_func(arr, exclude_values)


array([ 0,  1,  2,  3, -4, -5, -6, -7,  8,  9, 10])

## Задание 4
Напишите функцию, которая принимает одномерный положительный нампаевский массив и минимум 4 раза выводит на экран (функция ``print``) этот же массив, в котором у элементов оставлена только целая часть. Все >= 4 раза функция должна использовать разные методы для извлечения целой части.

**Sample:**
```python
#Исходный вектор 
arr = np.random.uniform(low=10, high=100,size=20)
#Исходный вектор arr:
arr
[99.54707368 49.26237526 80.50565073 10.10998238 16.39378916 60.64633657
 25.49637622 31.06574268 94.48636127 73.41204506 75.70638137 70.43830146
 72.28607422 40.59319738 57.85802554 35.39428061 55.25707514 48.00020145
 38.39449991 26.19085017]
 
my_func(arr)

[99. 49. 80. 10. 16. 60. 25. 31. 94. 73. 75. 70. 72. 40. 57. 35. 55. 48.
 38. 26.]
[99. 49. 80. 10. 16. 60. 25. 31. 94. 73. 75. 70. 72. 40. 57. 35. 55. 48.
 38. 26.]
[99. 49. 80. 10. 16. 60. 25. 31. 94. 73. 75. 70. 72. 40. 57. 35. 55. 48.
 38. 26.]
[99 49 80 10 16 60 25 31 94 73 75 70 72 40 57 35 55 48 38 26]
[99. 49. 80. 10. 16. 60. 25. 31. 94. 73. 75. 70. 72. 40. 57. 35. 55. 48.
 38. 26.]
```

В примере 5 выводов print из функции, просто чтобы были :)

In [21]:
import math

def my_func(a: np.array):
    """Функция выводит массив положительных чисел, в котором у элементов оставлена только целая часть
    несколькими способами"""

    # cпособ 1
    print(np.trunc(a))

    # способ 2
    print([int(x) for x in a])

    # способ 3
    print([x//1 for x in a])

    # способ 4
    print([x - x%1 for x in a])

    # способ 5
    print([math.modf(x)[1] for x in a])

    return None



#Исходный вектор 
arr = np.random.uniform(low=10, high=100,size=20)
#Исходный вектор arr:
print (arr)
 
my_func(arr)

[67.88648036 80.23712038 87.67605248 47.69609745 60.36806617 97.9391516
 16.14014527 37.98116248 75.00860521 21.15314057 63.17736446 24.15352664
 81.43749265 25.66070713 90.08209371 71.4315223  33.09990528 49.2024744
 34.47920244 80.77696735]
[67. 80. 87. 47. 60. 97. 16. 37. 75. 21. 63. 24. 81. 25. 90. 71. 33. 49.
 34. 80.]
[67, 80, 87, 47, 60, 97, 16, 37, 75, 21, 63, 24, 81, 25, 90, 71, 33, 49, 34, 80]
[67.0, 80.0, 87.0, 47.0, 60.0, 97.0, 16.0, 37.0, 75.0, 21.0, 63.0, 24.0, 81.0, 25.0, 90.0, 71.0, 33.0, 49.0, 34.0, 80.0]
[67.0, 80.0, 87.0, 47.0, 60.0, 97.0, 16.0, 37.0, 75.0, 21.0, 63.0, 24.0, 81.0, 25.0, 90.0, 71.0, 33.0, 49.0, 34.0, 80.0]
[67.0, 80.0, 87.0, 47.0, 60.0, 97.0, 16.0, 37.0, 75.0, 21.0, 63.0, 24.0, 81.0, 25.0, 90.0, 71.0, 33.0, 49.0, 34.0, 80.0]


## Задание 5
Напишите функцию, которая принимает два числа - ``N`` и ``p``. ``N`` - это размер стороны квадратной матрицы np.zeros, ``p`` - это количество индексов в матрице (p не превышает N*N), которые нужно заменить на случайные числа. Индексы и числа нужно выбирать случайно. Функция должна возвращать квадратную матрицу со стороной ``N``, в которой ровно ``p`` чисел заменены на случайные.

```python
my_func(N=6, p=4)

[[56.   0.   0.   0.   0.   0.]
 [  0.   0.   0.   0.   0.   0.]
 [99.   0.   0.   0.   0.   0.]
 [  0.   0.   0. 83.   0.   0.]
 [  0.   0.   0.   0.   0.   0.]
 [  0.   0.   0.   0.   1.   0.]]
```

In [22]:
def my_func(n: int, p: int) -> np.array:
    """функция строит матрицу размером n на n, заполненную нулями, и
    заменяет p случайных ячеек случайными числами"""

    if p > n*n:
        print ('Неверный ввод')
        return None # исключаем случай, когда случайных чисел больше, чем элементов в матрице.

    else:
        res = np.zeros((n*n)) # строим одномерный массив длиной n*n (после будет преобразован в нужный размер)
        
        indexes = np.random.permutation(np.arange((n*n)))[:p] # определяем индексы, в которых нужно менять значения (первые p значений от перемешенного массива чисел от 0 до n)

        for i in indexes:
            res[i] = np.random.randint(0,100) # присвиваем случайные значения по полученным индексам

        res = res.reshape((n,n)) # преобразовываем линейный массив в квадратную матрицу
    

        return res



my_func(n=6, p=4)

array([[ 0.,  0., 82.,  0.,  0.,  0.],
       [ 0.,  0.,  0.,  0.,  0.,  0.],
       [35.,  0.,  0.,  0.,  0.,  0.],
       [ 0.,  0., 87.,  0.,  0.,  0.],
       [ 0.,  0.,  0., 23.,  0.,  0.],
       [ 0.,  0.,  0.,  0.,  0.,  0.]])

# Pandas

In [23]:
import pandas as pd

# Задание 6

Любым удобным для вас способом создайте ``dataframe`` c именем ``my_data``, в котором две колонки c именами (``type`` - строки, ``value`` - числа с плавающей точкой) и четыре наблюдения в каждой колонке:

```
type value
B    18.

B    20.

C    10.

C    -10.
```

Какой размер созданного вами датафрейма? (Подумайте, что лучше использовать - информацию из метода df.info() или информацию из метода df.memory_usage()?)
1) Измените тип данных колонки type на category. Как изменился размер датафрейма?
2) Измените тип данных колонки value на менее жадный. Как изменился размер датафрейма?

In [24]:
from optparse import Values

my_data = pd.DataFrame(data = [['B', 18.],['B', 20.],['C', 10.],['C', -10.]], columns = ['type', 'value']) # создание датафрейма

my_data.info() #вывод информации о датафрейме
print('')
print(my_data.memory_usage()) # вывод информации о размере датафрейма

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4 entries, 0 to 3
Data columns (total 2 columns):
 #   Column  Non-Null Count  Dtype  
---  ------  --------------  -----  
 0   type    4 non-null      object 
 1   value   4 non-null      float64
dtypes: float64(1), object(1)
memory usage: 192.0+ bytes

Index    128
type      32
value     32
dtype: int64


In [25]:
my_data['type'] = my_data['type'].astype('category') # изменение типа данных в столбце type на category

my_data.info() #вывод информации о датафрейме
print('')
print(my_data.memory_usage()) # вывод информации о размере датафрейма

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4 entries, 0 to 3
Data columns (total 2 columns):
 #   Column  Non-Null Count  Dtype   
---  ------  --------------  -----   
 0   type    4 non-null      category
 1   value   4 non-null      float64 
dtypes: category(1), float64(1)
memory usage: 288.0 bytes

Index    128
type     128
value     32
dtype: int64


In [26]:
my_data['value'] = my_data['value'].astype('int') # изменение типа данных в столбце value на int

my_data.info() #вывод информации о датафрейме
print('')
print(my_data.memory_usage()) # вывод информации о размере датафрейма

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4 entries, 0 to 3
Data columns (total 2 columns):
 #   Column  Non-Null Count  Dtype   
---  ------  --------------  -----   
 0   type    4 non-null      category
 1   value   4 non-null      int32   
dtypes: category(1), int32(1)
memory usage: 272.0 bytes

Index    128
type     128
value     16
dtype: int64


### Вывод
Изначально созданный датафрейм имел размер 192 байта.
При изменении типа данных в столбце 'type' на 'category' его размер увеличился с 32 до 128 байт. Общий размер датафрейма стал 288 байт.
При изменении типа данных в столбце 'value' с 'float' на 'int' его размер уменьшился с 32 до 16 байт. Общий размер датафрейма стал 16 байт.

Выбор между функциями df.info() и df.memory_usage() должен обуславлтиваться тем, что важнее исследователю: размер датафрейма или размер его столбцов. В нашем случае оправдано использование обоих методов.

# Задание 7
Любым удобным для вас способом создайте ``dataframe`` c именем ``my_data``, в котором две колонки c именами (``type`` - строки, ``value`` - числа с плавающей точкой) и десять тысяч (10 000) наблюдений в каждой колонке. В колонке ``type`` должно быть 10 уникальных значений.

```python
my_data.shape
____________
(10000, 2)
```

Какой размер созданного вами датафрейма? (Подумайте, что лучше использовать - информацию из метода df.info() или информацию из метода df.memory_usage()?)
1) Измените тип данных колонки type на category. Как изменился размер датафрейма?
2) Измените тип данных колонки value на менее жадный. Как изменился размер датафрейма?

Во сколько раз уменьшился размер исходного датафрейма после выполнения этих двух операций?

In [27]:
my_data = pd.DataFrame() # создание пустого датафрейма

types = np.random.randint(10, size = ((1000)))

values = np.trunc(np.random.rand(1000)*100)

my_data['type'] = types
my_data['value'] = values

my_data.info()
print('')
print(my_data.memory_usage())

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1000 entries, 0 to 999
Data columns (total 2 columns):
 #   Column  Non-Null Count  Dtype  
---  ------  --------------  -----  
 0   type    1000 non-null   int32  
 1   value   1000 non-null   float64
dtypes: float64(1), int32(1)
memory usage: 11.8 KB

Index     128
type     4000
value    8000
dtype: int64


In [28]:
my_data['type'] = my_data['type'].astype('category') # изменение типа данных в столбце type на category

my_data.info() #вывод информации о датафрейме
print('')
print(my_data.memory_usage()) # вывод информации о размере датафрейма

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1000 entries, 0 to 999
Data columns (total 2 columns):
 #   Column  Non-Null Count  Dtype   
---  ------  --------------  -----   
 0   type    1000 non-null   category
 1   value   1000 non-null   float64 
dtypes: category(1), float64(1)
memory usage: 9.3 KB

Index     128
type     1380
value    8000
dtype: int64


In [29]:
my_data['value'] = my_data['value'].astype('int') # изменение типа данных в столбце value на int

my_data.info() #вывод информации о датафрейме
print('')
print(my_data.memory_usage()) # вывод информации о размере датафрейма

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1000 entries, 0 to 999
Data columns (total 2 columns):
 #   Column  Non-Null Count  Dtype   
---  ------  --------------  -----   
 0   type    1000 non-null   category
 1   value   1000 non-null   int32   
dtypes: category(1), int32(1)
memory usage: 5.4 KB

Index     128
type     1380
value    4000
dtype: int64


### Вывод
Изначально созданный датафрейм имел размер 11.8 kB.

При изменении типа данных в столбце 'type' на 'category' его размер уменьшился с 4000 до 1380 байт. Общий размер датафрейма стал 9.3 KB.
При изменении типа данных в столбце 'value' с 'float' на 'int' его размер уменьшился с 8000 до 4000 байт. Общий размер датафрейма стал 5.4 KB.

Обращает на себя внимание факт, что при большом наборе данных изменение типа на 'category' имеет значительный эффект, хотя в задаче 6 с 4 строками в датафрейме применение типа 'category' имело обратный эффект.

#### Для решения дальнейших задач мы будем использовать следующий датасет, загрузите его с помощью команды:

In [30]:
df = pd.read_csv('https://stepik.org/media/attachments/course/4852/my_stat.csv')

In [31]:
df

V1    V2 V3   V4
0     2  13.0  B  1.0
1    -1   7.0  B  1.0
2     0  11.0  A  0.0
3     2  11.0  A  2.0
4     0  10.0  B -1.0
..   ..   ... ..  ...
495   1   8.0  A -1.0
496   3   8.0  A  1.0
497  -1  11.0  B  2.0
498   1  11.0  A  2.0
499   0   8.0  B  1.0

[500 rows x 4 columns]

Примечание: все операции с исходным датафреймом должны быть ``inplace``. То есть исходный датафрейм в переменной ``df`` **должен** измениться

# Задание 8

В переменную с именем subset_1 сохраните только первые 19 строк и только 1 и 3 колонку.

В переменную с именем subset_2 сохраните все строки кроме 2, 3 и 18 и только 2 и 4 колонку.

In [37]:
df.info()
df.head()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 500 entries, 0 to 499
Data columns (total 4 columns):
 #   Column  Non-Null Count  Dtype  
---  ------  --------------  -----  
 0   V1      500 non-null    int64  
 1   V2      500 non-null    float64
 2   V3      500 non-null    object 
 3   V4      500 non-null    float64
dtypes: float64(2), int64(1), object(1)
memory usage: 15.8+ KB


V1    V2 V3   V4
0   2  13.0  B  1.0
1  -1   7.0  B  1.0
2   0  11.0  A  0.0
3   2  11.0  A  2.0
4   0  10.0  B -1.0

In [50]:
subset_1 = df.iloc[0:19,[0,2]] # сохранение первых 19 строк, первой и третьей строки
display(subset_1)


subset_2 = df.iloc[~df.index.isin([1,2,17]),[1,3]] # сохранение 2 и 4 столбца, все строки кроме 2,3,18.
subset_2

V1 V3
0    2  B
1   -1  B
2    0  A
3    2  A
4    0  B
5    2  A
6    0  A
7    1  A
8    0  B
9   -1  A
10   1  A
11  -2  A
12   0  B
13   1  A
14  -1  A
15   2  B
16   0  B
17   0  A
18   0  A

V2   V4
0    13.0  1.0
3    11.0  2.0
4    10.0 -1.0
5    11.0  0.0
6     9.0  2.0
..    ...  ...
495   8.0 -1.0
496   8.0  1.0
497  11.0  2.0
498  11.0  2.0
499   8.0  1.0

[497 rows x 2 columns]

# Задание 9

В переменную subset_3 сохраните только те наблюдения, у которых значения переменной V1 строго больше 0, и значение переменной V3  равняется 'A'.

В переменную  subset_4  сохраните только те наблюдения, у которых значения переменной V2  не равняются 10, или значения переменной V4 больше или равно 1.

In [53]:
subset_3 = df.loc[(df['V1'] > 0) & (df['V3'] == 'A')] # сохранение наблюдений, где V1 > 0, а V3 = 'А'
display(subset_3)
subset_4 = df.loc[(df['V2'] != 10) & (df['V4'] >= 1)] # сохранение наблюдений, где V2 не равно 10, а V4 больше или равно 1
subset_4

V1    V2 V3   V4
3     2  11.0  A  2.0
5     2  11.0  A  0.0
7     1   9.0  A  1.0
10    1  10.0  A  1.0
13    1  12.0  A  3.0
..   ..   ... ..  ...
491   4   7.0  A  2.0
494   1  11.0  A  1.0
495   1   8.0  A -1.0
496   3   8.0  A  1.0
498   1  11.0  A  2.0

[88 rows x 4 columns]

V1    V2 V3   V4
0     2  13.0  B  1.0
1    -1   7.0  B  1.0
3     2  11.0  A  2.0
6     0   9.0  A  2.0
7     1   9.0  A  1.0
..   ..   ... ..  ...
494   1  11.0  A  1.0
496   3   8.0  A  1.0
497  -1  11.0  B  2.0
498   1  11.0  A  2.0
499   0   8.0  B  1.0

[214 rows x 4 columns]

# Задание 10
Добавьте следующие колонки в df:

1) V5 = V1 + V4
2) V6 = **натуральный** логарифм переменной V2

In [58]:
df['V5'] = df['V1'] + df['V4'] # создание столбца с суммой столбцов V1 и V4
df['V6'] = [math.log(x) for x in df['V2']] # создание столбца с натуральным логарифмом столбца V2

df.head()

V1    V2 V3   V4   V5        V6
0   2  13.0  B  1.0  3.0  2.564949
1  -1   7.0  B  1.0  0.0  1.945910
2   0  11.0  A  0.0  0.0  2.397895
3   2  11.0  A  2.0  4.0  2.397895
4   0  10.0  B -1.0 -1.0  2.302585

# Задание 11

Переименуйте колонки в df.

1) V1 -> session_value

2) V2 -> group

3) V3 -> time

4) V4 -> n_users

In [59]:
df.rename(columns = {'V1':'session_value', 'V2':'group','V3':'time','V4':'n_users'}, inplace = True )
# переименование столбцов функцией rename

df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 500 entries, 0 to 499
Data columns (total 6 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   session_value  500 non-null    int64  
 1   group          500 non-null    float64
 2   time           500 non-null    object 
 3   n_users        500 non-null    float64
 4   V5             500 non-null    float64
 5   V6             500 non-null    float64
dtypes: float64(4), int64(1), object(1)
memory usage: 23.6+ KB


# Задание 12
В датафрейме ``my_stat`` в колонке ``session_value`` замените все пропущенные значения на нули.

В колонке ``n_users`` замените все отрицательные значения на медианное значение переменной ``n_users`` (без учета отрицательных значений, разумеется).

In [67]:
my_stat = pd.read_csv('https://stepik.org/media/attachments/course/4852/my_stat_1.csv')

my_stat.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10 entries, 0 to 9
Data columns (total 4 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   session_value  8 non-null      float64
 1   time           10 non-null     int64  
 2   group          10 non-null     object 
 3   n_users        10 non-null     int64  
dtypes: float64(1), int64(2), object(1)
memory usage: 448.0+ bytes


In [70]:
my_stat['session_value'].fillna(0, inplace=True) # заполнение пропусков в столбце 'session_value' на нули

my_stat[my_stat.n_users < 0] = my_stat[(my_stat.n_users >= 0)].n_users.median() # замена отрицательных значений в столбце n_users на медианное

my_stat

session_value  time group  n_users
0            0.0    12     B        9
1            5.0     5   5.0        5
2            1.0     8     A        1
3            2.0     9     B        3
4            2.0    10     B        9
5            3.0     8     A        8
6            0.0    11     B        5
7            0.0     9     A        4
8            5.0     5   5.0        5
9            5.0     5   5.0        5